File: windows/DoNotType.Core/OggOpusWriter.cs
namespace DoNotType.Core;

/// <summary>
/// Packages raw Opus packets into an Ogg stream.
/// </summary>
/// <remarks>
/// A port of <c>OggOpusWriter.swift</c>, checked byte-for-byte against the same reference stream
/// the Kotlin port passes (<c>eval/conformance/ogg-reference.bin</c>). The container is the part
/// most likely to be subtly wrong in a way that still produces a file decoders mostly accept — the
/// Swift version shipped two such bugs, one of which only surfaced as a message at the very end of
/// ffprobe's output.
///
/// Two mistakes worth not repeating, since both produced files that looked fine. Ending the stream
/// with an empty page to carry the end-of-stream flag is a zero-length Opus packet, which decoders
/// read past and then reject. And one packet per page costs 28 bytes of header per 20 ms frame —
/// 1.4 kB/s, nearly as much again as a 16 kbps stream.
/// </remarks>
public sealed class OggOpusWriter
{
    /// <summary>Opus reports timestamps at 48 kHz whatever the capture rate. A granule position in
    /// the wrong clock makes the file play at the wrong speed, or be rejected outright.</summary>
    public const int OpusClockRate = 48_000;

    /// <summary>50 × 20 ms per page. Larger pages save nothing measurable and delay nothing, since
    /// the whole file is written before any of it is sent.</summary>
    private const int PacketsPerPage = 50;

    private readonly int _sampleRate;
    private readonly int _channels;
    private readonly int _preSkip;
    private readonly uint _serial;

    private readonly MemoryStream _output = new();
    private readonly List<byte[]> _pending = [];
    private uint _sequence;
    private long _granule;
    private long _pendingGranule;

    public OggOpusWriter(
        int sampleRate = 16_000, int channels = 1, int preSkip = 312, uint serial = 0x646E7401)
    {
        _sampleRate = sampleRate;
        _channels = channels;
        _preSkip = preSkip;
        _serial = serial;
    }

    /// <summary>Writes the two mandatory headers. Must be called before any audio.</summary>
    public void Begin()
    {
        WritePage([OpusHead()], headerType: 0x02, granule: 0);
        WritePage([OpusTags()], headerType: 0x00, granule: 0);
    }

    /// <summary>Appends one encoded Opus packet.</summary>
    /// <param name="frameCount">
    /// Decoded samples the packet represents at the source rate. Converted to the 48 kHz Opus clock
    /// internally, because that is what a granule position must be expressed in.
    /// </param>
    public void Append(byte[] packet, int frameCount)
    {
        _granule += (long)frameCount * OpusClockRate / Math.Max(_sampleRate, 1);
        _pending.Add(packet);
        _pendingGranule = _granule;

        if (_pending.Count >= PacketsPerPage)
        {
            FlushPending(endOfStream: false);
        }
    }

    /// <summary>Finishes the stream and returns the complete file.</summary>
    public byte[] Finish()
    {
        FlushPending(endOfStream: true);
        return _output.ToArray();
    }

    private void FlushPending(bool endOfStream)
    {
        if (_pending.Count == 0)
        {
            return;
        }
        WritePage(_pending.ToArray(), endOfStream ? (byte)0x04 : (byte)0x00, _pendingGranule);
        _pending.Clear();
    }

    private byte[] OpusHead()
    {
        using var stream = new MemoryStream();
        using var writer = new BinaryWriter(stream);
        writer.Write("OpusHead"u8);
        writer.Write((byte)1); // version
        writer.Write((byte)_channels);
        writer.Write((ushort)_preSkip);
        writer.Write((uint)_sampleRate); // informational only
        writer.Write((ushort)0); // output gain
        writer.Write((byte)0); // channel mapping family
        writer.Flush();
        return stream.ToArray();
    }

    private static byte[] OpusTags()
    {
        var vendor = "DoNotType"u8.ToArray();
        using var stream = new MemoryStream();
        using var writer = new BinaryWriter(stream);
        writer.Write("OpusTags"u8);
        writer.Write((uint)vendor.Length);
        writer.Write(vendor);
        writer.Write((uint)0); // no user comments
        writer.Flush();
        return stream.ToArray();
    }

    /// <summary>Builds one Ogg page.</summary>
    /// <remarks>
    /// Packets larger than 255×255 bytes would have to span pages. A 20 ms Opus frame at any sane
    /// bitrate is a few hundred bytes, so the case cannot arise and is not handled — silently
    /// truncating would be worse than not supporting it.
    /// </remarks>
    private void WritePage(byte[][] payloads, byte headerType, long granule)
    {
        var segments = new List<byte>();
        var body = new List<byte>();

        foreach (var payload in payloads)
        {
            var remaining = payload.Length;
            while (remaining >= 255)
            {
                segments.Add(255);
                remaining -= 255;
            }
            segments.Add((byte)remaining);
            body.AddRange(payload);
        }

        using var stream = new MemoryStream();
        using (var writer = new BinaryWriter(stream, System.Text.Encoding.UTF8, leaveOpen: true))
        {
            writer.Write("OggS"u8);
            writer.Write((byte)0); // stream structure version
            writer.Write(headerType);
            writer.Write(granule);
            writer.Write(_serial);
            writer.Write(_sequence);
            writer.Write((uint)0); // CRC placeholder
            writer.Write((byte)segments.Count);
            writer.Write(segments.ToArray());
            writer.Write(body.ToArray());
        }

        _sequence++;

        // The checksum covers the whole page with its own field zeroed, then replaces it — which is
        // why it cannot be written above.
        var page = stream.ToArray();
        var checksum = Crc32(page);
        page[22] = (byte)(checksum & 0xFF);
        page[23] = (byte)((checksum >> 8) & 0xFF);
        page[24] = (byte)((checksum >> 16) & 0xFF);
        page[25] = (byte)((checksum >> 24) & 0xFF);

        _output.Write(page, 0, page.Length);
    }

    /// <summary>
    /// Ogg's CRC-32: polynomial 0x04C11DB7, no reflection, zero seed — <em>not</em> zip's CRC-32,
    /// which produces a wrong-but-plausible value that decoders reject.
    /// </summary>
    public static uint Crc32(ReadOnlySpan<byte> data)
    {
        uint crc = 0;
        foreach (var value in data)
        {
            crc ^= (uint)value << 24;
            for (var bit = 0; bit < 8; bit++)
            {
                crc = (crc & 0x80000000) != 0 ? (crc << 1) ^ 0x04C11DB7 : crc << 1;
            }
        }
        return crc;
    }
}

File: android/app/src/main/kotlin/app/donottype/core/OggOpusWriter.kt
package app.donottype.core

import java.io.ByteArrayOutputStream

/**
 * Packages raw Opus packets into an Ogg stream.
 *
 * A port of `OggOpusWriter.swift`, deliberately close to it. `MediaCodec` produces Opus packets and
 * `MediaMuxer` can write Ogg, but going through the muxer means a temporary file and a second
 * read; writing the framing here means one implementation of the container across four platforms,
 * and it is the part most likely to be subtly wrong in a way that still looks like a valid file.
 *
 * Two mistakes from the Swift version are worth not repeating, because both produced files that
 * looked fine. Ending the stream with an empty page to carry the end-of-stream flag is a
 * zero-length Opus packet, which decoders read past and then reject. And one packet per page costs
 * 28 bytes of header per 20 ms frame -- 1.4 kB/s, nearly as much again as a 16 kbps stream.
 */
class OggOpusWriter(
    private val sampleRate: Int = 16_000,
    private val channels: Int = 1,
    /** Encoder delay to discard on playback, in 48 kHz samples. */
    private val preSkip: Int = 312,
    private val serial: Int = 0x646E7401,
) {
    companion object {
        /** Opus reports timestamps at 48 kHz whatever the capture rate. A granule position in the
         *  wrong clock makes the file play at the wrong speed, or be rejected outright. */
        const val OPUS_CLOCK_RATE = 48_000

        /** 50 × 20 ms per page. Larger pages save nothing measurable and delay nothing, since the
         *  whole file is written before any of it is sent. */
        private const val PACKETS_PER_PAGE = 50

        /**
         * Ogg's CRC-32: polynomial 0x04C11DB7, no reflection, zero seed -- *not* zip's CRC-32,
         * which produces a wrong-but-plausible value that decoders reject.
         */
        fun crc32(data: ByteArray): Int {
            var crc = 0
            for (byte in data) {
                crc = crc xor ((byte.toInt() and 0xFF) shl 24)
                repeat(8) {
                    crc = if (crc and 0x80000000.toInt() != 0) {
                        (crc shl 1) xor 0x04C11DB7
                    } else {
                        crc shl 1
                    }
                }
            }
            return crc
        }
    }

    private val output = ByteArrayOutputStream()
    private val pending = mutableListOf<ByteArray>()
    private var sequence = 0
    private var granule = 0L
    private var pendingGranule = 0L

    /** Writes the two mandatory headers. Must be called before any audio. */
    fun begin() {
        output.write(page(listOf(opusHead()), headerType = 0x02, granule = 0))
        output.write(page(listOf(opusTags()), headerType = 0x00, granule = 0))
    }

    /**
     * Appends one encoded Opus packet.
     *
     * @param frameCount decoded samples the packet represents at the source rate; converted to the
     *   48 kHz Opus clock internally, because that is what a granule position must be expressed in.
     */
    fun append(packet: ByteArray, frameCount: Int) {
        granule += frameCount.toLong() * OPUS_CLOCK_RATE / maxOf(sampleRate, 1)
        pending += packet
        pendingGranule = granule
        if (pending.size >= PACKETS_PER_PAGE) flushPending(endOfStream = false)
    }

    /** Finishes the stream and returns the complete file. */
    fun finish(): ByteArray {
        flushPending(endOfStream = true)
        return output.toByteArray()
    }

    private fun flushPending(endOfStream: Boolean) {
        if (pending.isEmpty()) return
        output.write(page(pending.toList(), if (endOfStream) 0x04 else 0x00, pendingGranule))
        pending.clear()
    }

    private fun opusHead(): ByteArray {
        val out = ByteArrayOutputStream()
        out.write("OpusHead".toByteArray())
        out.write(1) // version
        out.write(channels)
        out.write(int16(preSkip))
        out.write(int32(sampleRate)) // informational only
        out.write(int16(0)) // output gain
        out.write(0) // channel mapping family
        return out.toByteArray()
    }

    private fun opusTags(): ByteArray {
        val vendor = "DoNotType".toByteArray()
        val out = ByteArrayOutputStream()
        out.write("OpusTags".toByteArray())
        out.write(int32(vendor.size))
        out.write(vendor)
        out.write(int32(0)) // no user comments
        return out.toByteArray()
    }

    /**
     * Builds one Ogg page.
     *
     * Packets larger than 255×255 bytes would have to span pages. A 20 ms Opus frame at any sane
     * bitrate is a few hundred bytes, so the case cannot arise and is not handled -- silently
     * truncating would be worse than not supporting it.
     */
    private fun page(payloads: List<ByteArray>, headerType: Int, granule: Long): ByteArray {
        val segments = ByteArrayOutputStream()
        val body = ByteArrayOutputStream()

        for (payload in payloads) {
            var remaining = payload.size
            while (remaining >= 255) {
                segments.write(255)
                remaining -= 255
            }
            segments.write(remaining)
            body.write(payload)
        }

        val segmentTable = segments.toByteArray()
        val page = ByteArrayOutputStream()
        page.write("OggS".toByteArray())
        page.write(0) // stream structure version
        page.write(headerType)
        page.write(int64(granule))
        page.write(int32(serial))
        page.write(int32(sequence))
        page.write(int32(0)) // CRC placeholder
        page.write(segmentTable.size)
        page.write(segmentTable)
        page.write(body.toByteArray())

        sequence++

        // The checksum covers the whole page with its own field zeroed, then replaces it -- which
        // is why it cannot be filled in above.
        val bytes = page.toByteArray()
        val checksum = crc32(bytes)
        bytes[22] = (checksum and 0xFF).toByte()
        bytes[23] = ((checksum shr 8) and 0xFF).toByte()
        bytes[24] = ((checksum shr 16) and 0xFF).toByte()
        bytes[25] = ((checksum shr 24) and 0xFF).toByte()
        return bytes
    }

    private fun int16(value: Int) = byteArrayOf(
        (value and 0xFF).toByte(),
        ((value shr 8) and 0xFF).toByte(),
    )

    private fun int32(value: Int) = byteArrayOf(
        (value and 0xFF).toByte(),
        ((value shr 8) and 0xFF).toByte(),
        ((value shr 16) and 0xFF).toByte(),
        ((value shr 24) and 0xFF).toByte(),
    )

    private fun int64(value: Long) = ByteArray(8) { ((value shr (it * 8)) and 0xFF).toByte() }
}
